File: monitor/internal/store/store.go
// Package store persists incidents to SQLite — the single file that acts
// as the contract between the monitor (writer), the diagnosis agent
// (Phase 3, reader/updater), and the dashboard (Phase 5, reader).
//
// Driver: modernc.org/sqlite (pure Go). See docs/design-decisions.md for
// why it was picked over the cgo-based alternative.
package store

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	_ "modernc.org/sqlite"

	"github.com/1816x/self-healing-agent/monitor/internal/detect"
)

// schemaVersion is the current PRAGMA user_version. Open() walks a
// hand-built DB from any earlier version up to this one — the honest
// answer to "how does a file-based store evolve its schema without a
// migration framework" for a project this size.
const schemaVersion = 4

const schemaV1 = `
CREATE TABLE IF NOT EXISTS incidents (
	id           INTEGER PRIMARY KEY AUTOINCREMENT,
	created_at   TEXT NOT NULL,
	kind         TEXT NOT NULL,
	status       TEXT NOT NULL DEFAULT 'detected',
	window_start TEXT NOT NULL,
	window_end   TEXT NOT NULL,
	evidence     TEXT NOT NULL
);`

// v2 adds what Phase 2's dedup/cooldown needs: a stable identity to merge
// repeat firings against (dedup_key), when an open incident was last
// touched (updated_at), and how many times it's re-fired (occurrences).
// F1 shipped without these, so a real DB from that phase has to migrate
// forward, not just get re-created.
var schemaV2Statements = []string{
	`ALTER TABLE incidents ADD COLUMN dedup_key TEXT NOT NULL DEFAULT ''`,
	`ALTER TABLE incidents ADD COLUMN updated_at TEXT NOT NULL DEFAULT ''`,
	`ALTER TABLE incidents ADD COLUMN occurrences INTEGER NOT NULL DEFAULT 1`,
	`UPDATE incidents SET dedup_key = kind WHERE dedup_key = ''`,
	`UPDATE incidents SET updated_at = window_end WHERE updated_at = ''`,
}

// v3 adds the columns the Phase 3 diagnosis agent writes back. The Go
// monitor never writes them — it only ever inserts and merges incidents —
// but the DDL lives here anyway: this package already owns the migration
// ladder, and one language owning all schema beats two languages agreeing
// on it. The Python agent asserts user_version >= 3 instead of carrying a
// duplicate copy of these statements (see docs/design-decisions.md).
var schemaV3Statements = []string{
	`ALTER TABLE incidents ADD COLUMN diagnosis TEXT NOT NULL DEFAULT ''`,
	`ALTER TABLE incidents ADD COLUMN diagnosed_at TEXT NOT NULL DEFAULT ''`,
	`ALTER TABLE incidents ADD COLUMN proposed_fix TEXT NOT NULL DEFAULT ''`,
}

// v4 adds what Phase 4 needs to record the outcome of the validation gate
// and the pull request it opens. validation holds the gate's evidence (did
// the diff apply, were the tests red before and green after) and is kept
// separate from diagnosis: one is what the model claimed, the other is
// what this machine independently checked, and collapsing them would make
// a proposal indistinguishable from a verified fix.
var schemaV4Statements = []string{
	`ALTER TABLE incidents ADD COLUMN validation TEXT NOT NULL DEFAULT ''`,
	`ALTER TABLE incidents ADD COLUMN pr_url TEXT NOT NULL DEFAULT ''`,
}

// Incident lifecycle. The monitor only ever produces StatusDetected; every
// later transition is written by the Phase 3 agent (and Phase 4's PR step).
// They're declared here so both sides share one vocabulary rather than
// trading bare strings.
const (
	StatusDetected = "detected"
	// StatusDiagnosing is claimed atomically, so two agent processes can't
	// pick up the same incident.
	StatusDiagnosing = "diagnosing"
	StatusDiagnosed  = "diagnosed"
	// StatusFixProposed means a diff was recorded. Phase 4 adds the
	// validation gate and the PR; Phase 3 only records.
	StatusFixProposed = "fix_proposed"
	// StatusDiagnosisFailed covers a tool-loop error or a blown iteration
	// cap; StatusDiagnosisRefused covers stop_reason == "refusal" from the
	// model, which is a distinct outcome worth not conflating with a bug.
	StatusDiagnosisFailed  = "diagnosis_failed"
	StatusDiagnosisRefused = "diagnosis_refused"
	// StatusFixValidated means the proposed diff applied cleanly and the
	// demo app's tests passed with it. It is deliberately distinct from
	// StatusPROpened: validation is local and always runs, opening a PR is
	// a separate, opt-in, outward-facing act.
	StatusFixValidated = "fix_validated"
	// StatusFixFailed means the gate rejected the diff — it didn't apply,
	// or the tests didn't pass with it. Terminal: no PR is opened.
	StatusFixFailed = "fix_failed"
	StatusPROpened  = "pr_opened"
)

// UpsertOutcome says what UpsertIncident did with a firing. Three outcomes,
// because a sustained failure produces far more firings than incidents and
// the daemon's log should distinguish "new problem" from "still happening"
// from "already being worked on".
type UpsertOutcome int

const (
	// Inserted: this firing opened a new incident row.
	Inserted UpsertOutcome = iota
	// Merged: it extended an existing incident still waiting to be picked up.
	Merged
	// Suppressed: an incident for this condition already exists and has been
	// claimed downstream. Nothing was written.
	Suppressed
)

func (o UpsertOutcome) String() string {
	switch o {
	case Merged:
		return "ongoing"
	case Suppressed:
		return "already claimed"
	default:
		return "new"
	}
}

type Store struct {
	db *sql.DB
	// SQLite allows one writer at a time. Phase 2 adds a second goroutine
	// (the metrics scrape loop) writing incidents alongside the existing
	// log-tailer loop; a mutex around every write is simpler and more
	// certain than relying on busy_timeout retries to paper over
	// "database is locked" under concurrent access.
	mu sync.Mutex
}

// evidence is the JSON blob stored per incident: everything the Phase 3
// agent needs to start a diagnosis without re-reading the whole log.
type evidence struct {
	Summary string             `json:"summary"`
	Metrics map[string]float64 `json:"metrics"`
	Routes  map[string]int     `json:"routes,omitempty"`
	Samples []string           `json:"samples"`
}

// maxStoredSamples caps evidence growth across repeated merges of a
// still-ongoing incident — distinct from each detector's own per-firing
// sample cap (see detect.maxSamples), which bounds a single Incident.
const maxStoredSamples = 10

// busyTimeout is how long SQLite waits for a lock held by another
// *process* before giving up. The mutex above only serializes this
// process's own goroutines; it says nothing about the diagnosis agent or
// the dashboard, which open the same file whenever they like. Without
// this, a reader holding a shared lock for a few milliseconds is enough
// to make the monitor's migration fail with SQLITE_BUSY and kill it at
// startup — which is exactly what the agent's test suite reproduced once
// the v4 migration widened the window.
const busyTimeout = 5 * time.Second

func Open(path string) (*Store, error) {
	// Encoded as a DSN pragma rather than a `PRAGMA busy_timeout` after
	// connecting: database/sql pools connections and hands out new ones on
	// demand, so a pragma run once on one connection would not apply to
	// the others. The DSN applies to every connection the pool opens.
	dsn := fmt.Sprintf("file:%s?_pragma=busy_timeout(%d)", path, busyTimeout.Milliseconds())
	db, err := sql.Open("sqlite", dsn)
	if err != nil {
		return nil, fmt.Errorf("open %s: %w", path, err)
	}
	if err := migrate(db); err != nil {
		db.Close()
		return nil, fmt.Errorf("migrate schema: %w", err)
	}
	return &Store{db: db}, nil
}

func migrate(db *sql.DB) error {
	var version int
	if err := db.QueryRow(`PRAGMA user_version`).Scan(&version); err != nil {
		return fmt.Errorf("read user_version: %w", err)
	}

	if version < 1 {
		if _, err := db.Exec(schemaV1); err != nil {
			return fmt.Errorf("apply v1 schema: %w", err)
		}
		version = 1
	}
	if version < 2 {
		for _, stmt := range schemaV2Statements {
			if _, err := db.Exec(stmt); err != nil {
				return fmt.Errorf("apply v2 migration (%s): %w", stmt, err)
			}
		}
		version = 2
	}
	if version < 3 {
		for _, stmt := range schemaV3Statements {
			if _, err := db.Exec(stmt); err != nil {
				return fmt.Errorf("apply v3 migration (%s): %w", stmt, err)
			}
		}
		version = 3
	}
	if version < 4 {
		for _, stmt := range schemaV4Statements {
			if _, err := db.Exec(stmt); err != nil {
				return fmt.Errorf("apply v4 migration (%s): %w", stmt, err)
			}
		}
		version = 4
	}

	// PRAGMA doesn't support bound parameters; version is our own int,
	// never user input, so this is safe to format directly.
	if _, err := db.Exec(fmt.Sprintf(`PRAGMA user_version = %d`, version)); err != nil {
		return fmt.Errorf("set user_version: %w", err)
	}
	return nil
}

func (s *Store) Close() error {
	return s.db.Close()
}

// execer is satisfied by both *sql.DB and *sql.Tx, so insert() works
// whether or not the caller is inside a transaction.
type execer interface {
	Exec(query string, args ...any) (sql.Result, error)
}

// InsertIncident stores a fired incident with status 'detected' and
// returns its row id, unconditionally — no dedup against any existing
// open incident. Kept for callers (and tests) that want an unconditional
// insert; the daemon itself uses UpsertIncident.
func (s *Store) InsertIncident(incident *detect.Incident) (int64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return insert(s.db, incident, 1)
}

func insert(x execer, incident *detect.Incident, occurrences int) (int64, error) {
	blob, err := json.Marshal(evidence{
		Summary: incident.Summary,
		Metrics: incident.Metrics,
		Routes:  incident.Routes,
		Samples: incident.Samples,
	})
	if err != nil {
		return 0, fmt.Errorf("marshal evidence: %w", err)
	}

	windowEnd := incident.WindowEnd.UTC().Format(time.RFC3339)
	result, err := x.Exec(
		`INSERT INTO incidents (created_at, kind, dedup_key, window_start, window_end, evidence, updated_at, occurrences)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		// created_at is real wall-clock time (operational metadata: when
		// the monitor actually wrote the row). updated_at tracks the
		// incident's own event-time (WindowEnd), not wall-clock — dedup
		// compares it against a later incident's WindowStart, and mixing
		// event-time with wall-clock there breaks both production
		// correctness under log-delivery lag and deterministic testing.
		time.Now().UTC().Format(time.RFC3339),
		incident.Kind,
		incident.DedupKey,
		incident.WindowStart.UTC().Format(time.RFC3339),
		windowEnd,
		string(blob),
		windowEnd,
		occurrences,
	)
	if err != nil {
		return 0, fmt.Errorf("insert incident: %w", err)
	}
	return result.LastInsertId()
}

// UpsertIncident merges a fresh firing into the still-open incident with
// the same DedupKey if one was updated within dedupWindow of this
// firing's start; otherwise it inserts a new row. Merging replaces the
// evidence with the incoming (latest) snapshot of the condition, extends
// window_end, appends samples up to maxStoredSamples, and increments
// occurrences — it does not attempt to numerically combine old and new
// Metrics, which would need per-metric semantics (sum vs. max vs.
// latest) this store has no way to know. "Latest state plus an
// occurrence count" is how most real incident/alerting systems show an
// ongoing condition, so the simplification is a reasonable one, not a
// missing feature.
//
// This is what turns a flapping signal into one incident row instead of
// one per detector reset — the gap the Phase 1 live run surfaced.
func (s *Store) UpsertIncident(incident *detect.Incident, dedupWindow time.Duration) (id int64, outcome UpsertOutcome, err error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	tx, err := s.db.Begin()
	if err != nil {
		return 0, Inserted, fmt.Errorf("begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck // no-op once committed

	var existingID int64
	var existingStatus, existingWindowEnd, existingEvidence string
	var existingOccurrences int
	// Deliberately not filtered by status. Restricting this to `detected`
	// rows made a claimed incident invisible here, so a still-firing
	// condition opened a brand new incident every detector cycle once the
	// agent started picking them up — the exact "one incident, not fifty"
	// property Phase 2 exists to provide.
	row := tx.QueryRow(
		`SELECT id, status, window_end, evidence, occurrences FROM incidents
		 WHERE dedup_key = ?
		 ORDER BY id DESC LIMIT 1`,
		incident.DedupKey,
	)
	scanErr := row.Scan(&existingID, &existingStatus, &existingWindowEnd, &existingEvidence, &existingOccurrences)

	switch {
	case scanErr == sql.ErrNoRows:
		newID, insertErr := insert(tx, incident, 1)
		if insertErr != nil {
			return 0, Inserted, insertErr
		}
		if err := tx.Commit(); err != nil {
			return 0, Inserted, fmt.Errorf("commit: %w", err)
		}
		return newID, Inserted, nil

	case scanErr != nil:
		return 0, Inserted, fmt.Errorf("query existing incident: %w", scanErr)
	}

	// window_end, not updated_at: window_end is always this store's event
	// time, written only here and in insert(). updated_at is shared with the
	// Python agent, which writes wall-clock into it on every state change —
	// comparing that against a detector's event clock mixes two timebases.
	lastFiring, parseErr := time.Parse(time.RFC3339, existingWindowEnd)
	if parseErr != nil || incident.WindowStart.Sub(lastFiring) > dedupWindow {
		newID, insertErr := insert(tx, incident, 1)
		if insertErr != nil {
			return 0, Inserted, insertErr
		}
		if err := tx.Commit(); err != nil {
			return 0, Inserted, fmt.Errorf("commit: %w", err)
		}
		return newID, Inserted, nil
	}

	// Within the window, but something downstream already owns this row.
	// Merging would rewrite evidence under an agent that is mid-diagnosis,
	// and inserting would duplicate a condition already being worked. So
	// this firing is dropped: the incident is on record and in progress.
	if existingStatus != StatusDetected {
		if err := tx.Commit(); err != nil {
			return 0, Suppressed, fmt.Errorf("commit: %w", err)
		}
		return existingID, Suppressed, nil
	}

	var prev evidence
	if err := json.Unmarshal([]byte(existingEvidence), &prev); err != nil {
		return 0, Merged, fmt.Errorf("unmarshal existing evidence: %w", err)
	}
	mergedEvidence := evidence{
		Summary: incident.Summary,
		Metrics: incident.Metrics,
		Routes:  incident.Routes,
		Samples: append(append([]string{}, prev.Samples...), incident.Samples...),
	}
	if len(mergedEvidence.Samples) > maxStoredSamples {
		mergedEvidence.Samples = mergedEvidence.Samples[len(mergedEvidence.Samples)-maxStoredSamples:]
	}
	blob, err := json.Marshal(mergedEvidence)
	if err != nil {
		return 0, Merged, fmt.Errorf("marshal merged evidence: %w", err)
	}

	windowEnd := incident.WindowEnd.UTC().Format(time.RFC3339)
	_, err = tx.Exec(
		`UPDATE incidents SET window_end = ?, updated_at = ?, evidence = ?, occurrences = ?
		 WHERE id = ?`,
		windowEnd,
		windowEnd, // event-time, not wall-clock — see the note in insert()
		string(blob),
		existingOccurrences+1,
		existingID,
	)
	if err != nil {
		return 0, Merged, fmt.Errorf("update incident: %w", err)
	}
	if err := tx.Commit(); err != nil {
		return 0, Merged, fmt.Errorf("commit: %w", err)
	}
	return existingID, Merged, nil
}

// CountByStatus is a small helper for tests and the daemon's own logging.
func (s *Store) CountByStatus(status string) (int, error) {
	var n int
	err := s.db.QueryRow(`SELECT COUNT(*) FROM incidents WHERE status = ?`, status).Scan(&n)
	return n, err
}

File: agent/diagnose/store.py
"""Read/write access to the shared incident store.

The Go monitor owns every schema change (it has the ``PRAGMA
user_version`` migration ladder). This module deliberately carries no
DDL: it asserts the version it needs and tells the operator to run the
monitor if the database is older. One language owning all schema beats
two languages agreeing on it — see ``docs/design-decisions.md``.
"""

from __future__ import annotations

import json
import sqlite3
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Self

# Must match monitor/internal/store/store.go's schemaVersion. The agent
# reads and writes columns v4 introduced, so anything older can't serve it.
REQUIRED_SCHEMA_VERSION = 4

STATUS_DETECTED = "detected"
STATUS_DIAGNOSING = "diagnosing"
STATUS_DIAGNOSED = "diagnosed"
STATUS_FIX_PROPOSED = "fix_proposed"
STATUS_DIAGNOSIS_FAILED = "diagnosis_failed"
STATUS_DIAGNOSIS_REFUSED = "diagnosis_refused"
# Phase 4. fix_validated and fix_failed are the two outcomes of the local
# validation gate; pr_opened is the separate, opt-in outward-facing step.
STATUS_FIX_VALIDATED = "fix_validated"
STATUS_FIX_FAILED = "fix_failed"
STATUS_PR_OPENED = "pr_opened"

# The default claim set: a normal run only ever picks up fresh incidents.
CLAIMABLE_DEFAULT = (STATUS_DETECTED,)

# What `--resume` adds. `fix_proposed` is a resting state only `--no-validate`
# produces: the model finished and its diff was stored, but nothing ever ran
# the validation gate over it. Before Phase 5 that row was unreachable, because
# every claim hardcoded `detected`.
CLAIMABLE_RESUME = (STATUS_DETECTED, STATUS_FIX_PROPOSED)

# How long an incident may sit in `diagnosing` before a resumed run assumes the
# process holding it died. Nothing renews a lease mid-run, so this has to
# exceed the longest plausible tool loop rather than the average one.
DEFAULT_LEASE_SECONDS = 900


class SchemaTooOldError(RuntimeError):
    """The database predates the columns this agent needs."""


@dataclass(frozen=True)
class Incident:
    """One incident row, as the agent needs to see it."""

    id: int
    kind: str
    dedup_key: str
    status: str
    window_start: str
    window_end: str
    occurrences: int
    evidence: dict

    def summary_line(self) -> str:
        """One-line description for the agent's opening prompt."""
        summary = self.evidence.get("summary") or self.kind
        return f"incident #{self.id} ({self.kind}): {summary}"


def _isoformat(moment: datetime) -> str:
    return moment.strftime("%Y-%m-%dT%H:%M:%SZ")


def _utcnow() -> str:
    return _isoformat(datetime.now(UTC))


class Store:
    def __init__(self, path: str):
        # isolation_level=None puts the connection in autocommit mode so an
        # explicit BEGIN IMMEDIATE in claim() means what it says instead of
        # racing sqlite3's implicit transaction handling.
        self._db = sqlite3.connect(path, isolation_level=None)
        self._db.row_factory = sqlite3.Row
        self._assert_schema_version()

    def _assert_schema_version(self) -> None:
        version = self._db.execute("PRAGMA user_version").fetchone()[0]
        if version < REQUIRED_SCHEMA_VERSION:
            raise SchemaTooOldError(
                f"incident store is at schema v{version}, agent needs "
                f"v{REQUIRED_SCHEMA_VERSION}. The monitor owns migrations — "
                f"run it once against this database to migrate it forward."
            )

    def close(self) -> None:
        self._db.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *_exc) -> None:
        self.close()

    # --- reading ---

    def get(self, incident_id: int) -> Incident | None:
        row = self._db.execute(
            """SELECT id, kind, dedup_key, status, window_start, window_end,
                      occurrences, evidence
               FROM incidents WHERE id = ?""",
            (incident_id,),
        ).fetchone()
        return _row_to_incident(row) if row else None

    def claim_next(self, statuses: Sequence[str] = CLAIMABLE_DEFAULT) -> Incident | None:
        """Claims the oldest incident in one of `statuses`, or returns None.

        The claim is a conditional UPDATE guarded by the current status, so
        two agent processes racing on the same store can't both pick up the
        same incident: exactly one UPDATE reports a changed row. Widening
        `statuses` doesn't weaken that — the UPDATE still names the exact set
        it selected against.
        """
        placeholders = ", ".join("?" for _ in statuses)
        self._db.execute("BEGIN IMMEDIATE")
        try:
            row = self._db.execute(
                f"""SELECT id FROM incidents WHERE status IN ({placeholders})
                    ORDER BY id LIMIT 1""",
                tuple(statuses),
            ).fetchone()
            if row is None:
                self._db.execute("COMMIT")
                return None

            incident_id = row["id"]
            cursor = self._db.execute(
                f"""UPDATE incidents SET status = ?, updated_at = ?
                    WHERE id = ? AND status IN ({placeholders})""",
                (STATUS_DIAGNOSING, _utcnow(), incident_id, *statuses),
            )
            if cursor.rowcount != 1:
                # Another process claimed it between the SELECT and UPDATE.
                self._db.execute("COMMIT")
                return None
            self._db.execute("COMMIT")
        except Exception:
            self._db.execute("ROLLBACK")
            raise

        return self.get(incident_id)

    def claim(
        self, incident_id: int, statuses: Sequence[str] = CLAIMABLE_DEFAULT
    ) -> Incident | None:
        """Claims one specific incident by id. Same guard as claim_next."""
        placeholders = ", ".join("?" for _ in statuses)
        cursor = self._db.execute(
            f"""UPDATE incidents SET status = ?, updated_at = ?
                WHERE id = ? AND status IN ({placeholders})""",
            (STATUS_DIAGNOSING, _utcnow(), incident_id, *statuses),
        )
        if cursor.rowcount != 1:
            return None
        return self.get(incident_id)

    def reclaim_stale_diagnosing(self, max_age_seconds: int) -> Incident | None:
        """Reclaims an incident abandoned in `diagnosing`, or returns None.

        A process killed between `claim()` and any writer leaves its row
        claimed forever: the CLI's broad excepts guarantee a *terminal* status
        on every path it survives, but SIGKILL is not a path it survives.
        This is the reaper for that case, and the first code to read
        `updated_at` back.

        That column has mixed semantics — Go writes event time, Python writes
        wall clock (`_set`). Reading it as wall clock is only sound because Go
        never writes `diagnosing`; it produces `detected` and nothing else. So
        every row this can match was last touched by `_set`.
        """
        cutoff = _isoformat(datetime.now(UTC) - timedelta(seconds=max_age_seconds))
        self._db.execute("BEGIN IMMEDIATE")
        try:
            row = self._db.execute(
                """SELECT id FROM incidents
                   WHERE status = ? AND updated_at != '' AND updated_at < ?
                   ORDER BY id LIMIT 1""",
                (STATUS_DIAGNOSING, cutoff),
            ).fetchone()
            if row is None:
                self._db.execute("COMMIT")
                return None

            incident_id = row["id"]
            # Re-assert both predicates: another resumed run may have taken it,
            # and the original process may have come back to life and written.
            cursor = self._db.execute(
                """UPDATE incidents SET updated_at = ?
                   WHERE id = ? AND status = ? AND updated_at < ?""",
                (_utcnow(), incident_id, STATUS_DIAGNOSING, cutoff),
            )
            if cursor.rowcount != 1:
                self._db.execute("COMMIT")
                return None
            self._db.execute("COMMIT")
        except Exception:
            self._db.execute("ROLLBACK")
            raise

        return self.get(incident_id)

    def read_fix_record(self, incident_id: int) -> tuple[dict, dict]:
        """Returns the stored `(diagnosis, proposed_fix)` for an incident.

        A resumed `fix_proposed` incident already has both — re-running the
        model to recover them would spend a real API call reproducing work
        that is sitting in the row. Kept as its own reader rather than added
        to `Incident`, which every other call site would then carry.

        Absent columns are the empty string, not NULL, so "" means "never
        written" and decodes to {}.
        """
        row = self._db.execute(
            "SELECT diagnosis, proposed_fix FROM incidents WHERE id = ?",
            (incident_id,),
        ).fetchone()
        if row is None:
            return {}, {}
        return _decode_blob(row["diagnosis"]), _decode_blob(row["proposed_fix"])

    def count_by_status(self, status: str) -> int:
        return self._db.execute(
            "SELECT COUNT(*) FROM incidents WHERE status = ?", (status,)
        ).fetchone()[0]

    # --- writing ---

    def write_diagnosis(self, incident_id: int, diagnosis: dict) -> None:
        """Records a completed diagnosis and moves the incident to `diagnosed`.

        `diagnosis` carries a `source` field ("live" / "replay" /
        "heuristic") so a consumer can never mistake a canned offline
        answer for a real model run.
        """
        self._set(
            incident_id,
            status=STATUS_DIAGNOSED,
            diagnosis=json.dumps(diagnosis),
            diagnosed_at=_utcnow(),
        )

    def write_proposed_fix(self, incident_id: int, diagnosis: dict, proposed_fix: dict) -> None:
        """Records a diagnosis plus a proposed diff, without applying it.

        Recording is still separate from validating: this only says the
        model produced a diff. Whether that diff applies and passes tests
        is `write_validated` / `write_validation_failed`, which run against
        the repository rather than trusting the proposal.
        """
        self._set(
            incident_id,
            status=STATUS_FIX_PROPOSED,
            diagnosis=json.dumps(diagnosis),
            diagnosed_at=_utcnow(),
            proposed_fix=json.dumps(proposed_fix),
        )

    def write_validated(self, incident_id: int, validation: dict) -> None:
        """The gate applied the diff and the tests passed with it.

        `validation` is stored in its own column rather than merged into
        `diagnosis` on purpose: one is the model's claim, the other is what
        this machine independently checked, and a dashboard must be able to
        tell a proposal from a verified fix.
        """
        self._set(
            incident_id,
            status=STATUS_FIX_VALIDATED,
            validation=json.dumps(validation),
        )

    def write_validation_failed(self, incident_id: int, validation: dict) -> None:
        """The diff didn't apply, or the tests didn't pass with it.

        Terminal — no PR is opened. The recorded evidence carries git's or
        pytest's own output so the failure is diagnosable without a rerun.
        """
        self._set(
            incident_id,
            status=STATUS_FIX_FAILED,
            validation=json.dumps(validation),
        )

    def write_pr_opened(self, incident_id: int, pr_url: str) -> None:
        self._set(incident_id, status=STATUS_PR_OPENED, pr_url=pr_url)

    def mark_failed(self, incident_id: int, reason: str) -> None:
        self._set(
            incident_id,
            status=STATUS_DIAGNOSIS_FAILED,
            diagnosis=json.dumps({"error": reason}),
            diagnosed_at=_utcnow(),
        )

    def mark_refused(self, incident_id: int, detail: dict) -> None:
        """The model declined the request (stop_reason == "refusal").

        Distinct from mark_failed: nothing is broken, and retrying the same
        prompt won't help — so this is recorded as its own outcome rather
        than as an error.
        """
        self._set(
            incident_id,
            status=STATUS_DIAGNOSIS_REFUSED,
            diagnosis=json.dumps({"refusal": detail}),
            diagnosed_at=_utcnow(),
        )

    def _set(self, incident_id: int, *, status: str, **columns) -> None:
        assignments = ", ".join(f"{name} = ?" for name in columns)
        values = list(columns.values())
        self._db.execute(
            f"UPDATE incidents SET status = ?, updated_at = ?, {assignments} WHERE id = ?",
            [status, _utcnow(), *values, incident_id],
        )


def _decode_blob(raw: str) -> dict:
    """Decodes one of the JSON columns, treating unset and corrupt alike.

    Every column added after schema v1 is `NOT NULL DEFAULT ''`, so "not
    written yet" arrives as the empty string and `json.loads` would raise on
    it. A caller asking for a fix that isn't there wants {}, not an exception.
    """
    if not raw:
        return {}
    try:
        decoded = json.loads(raw)
    except (json.JSONDecodeError, TypeError):
        return {}
    return decoded if isinstance(decoded, dict) else {}


def _row_to_incident(row: sqlite3.Row) -> Incident:
    try:
        evidence = json.loads(row["evidence"])
    except (json.JSONDecodeError, TypeError):
        # A corrupt evidence blob shouldn't make the incident undiagnosable —
        # the agent has tools to go read the logs itself.
        evidence = {}
    return Incident(
        id=row["id"],
        kind=row["kind"],
        dedup_key=row["dedup_key"],
        status=row["status"],
        window_start=row["window_start"],
        window_end=row["window_end"],
        occurrences=row["occurrences"],
        evidence=evidence,
    )
